### PR TITLE
Add constants for f16 and f128

### PR DESCRIPTION
**Pull request by tspiteri**
- Commit 1 adds associated constants for `f16`, excluding NaN and infinities as these are implemented using arithmetic for `f32` and `f64`.
- Commit 2 adds associated constants for `f128`, excluding NaN and infinities.
- Commit 3 adds constants in `std::f16::consts`.
- Commit 4 adds constants in `std::f128::consts`.

### CODE REVIEW

**Comment by fmease** [comment]
cc @tgross35

File: library/core/src/num/f128.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//! Constants for the `f128` quadruple-precision floating point type.
+//! constants for the `f128` quadruple-precision floating point type.
```
Comment (fmease) [inline library/core/src/num/f128.rs:1]:
```suggestion
//! Constants for the `f128` quadruple-precision floating point type.
```

**Comment by tspiteri** [comment]
Note that for constants gated under [`more_float_constants`](https://github.com/rust-lang/rust/issues/103883) for `f32` and `f64`, I used #103883 rather than #116909 as a tracking issue.

**Comment by tspiteri** [comment]
The test failure seems to be a panic in beta clippy.

```
  thread 'rustc' panicked at src/tools/clippy/clippy_lints/src/float_literal.rs:145:26:
  not implemented: f16_f128
```

**Review by tgross35 — CHANGES_REQUESTED** [review]
Hey, thank you for the help here!

I was holding off on these until we had the ability to actually run code that involves these types, which unfortunately we won't until I get `compiler_builtins` updated (well, code will run on x86 but nothing else so we can't actually merge anything). I was trying to add as little as possible until we get tests, but don't have any big problem with adding these.

One change needed - the literals we have aren't high enough precision for f128. I have a Julia script to generate longer ones using arbitrary precision math, https://github.com/tgross35/apfloat-consts. You can just copy the consts from output.rs.

**Comment by tgross35** [comment]
> The test failure seems to be a panic in beta clippy.
> 
> ```
>   thread 'rustc' panicked at src/tools/clippy/clippy_lints/src/float_literal.rs:145:26:
>   not implemented: f16_f128
> ```

This is the other reason I haven't added too much more, clippy and CTFE both have some `unimplemented!` scattered around that can't be fixed until math is working (compiler_builtins...). You might be able to `#[allow(clippy::all)]` on this module.

@rustbot label +F-f16_and_f128

**Comment by tspiteri** [comment]
Oops, I used 36 significant digits for `f128` constants. I stumbled because 36 decimal digits is enough for numbers which are representable (such as `EPSILON`, `MIN_POSITIVE`, `MAX`, etc.), but not for numbers that are not, where you may need many more digits especially if the exact decimal number is close to mid-way between two consecutive floating-point numbers.

I changed `f128` constants like `PI` to use 60 decimal significant digits, which is enough for all these constants. (I generated them with MPFR through the Rug crate, with rounding to the nearest just like the constants in the standard library for `f32` and `f64`.) For `f16` I just used the same literals that are used for `f32` and `f64`.

**Comment by tspiteri** [comment]
For the clippy warning, I think clippy is already fine in nightly, so instead of adding extra clippy allows, I think I might just wait for the next beta to be cut.

**Comment by tspiteri** [comment]
> For the clippy warning, I think clippy is already fine in nightly, so instead of adding extra clippy allows, I think I might just wait for the next beta to be cut.

I rebased now that the clippy tests in mingw-check are using 2024-04-29 instead of 2024-03-19, and the tests now pass.

**Comment by tgross35** [comment]
> I changed `f128` constants like `PI` to use 60 decimal significant digits, which is enough for all these constants. (I generated them with MPFR through the Rug crate, with rounding to the nearest just like the constants in the standard library for `f32` and `f64`.) For `f16` I just used the same literals that are used for `f32` and `f64`.

Could you post the code for this so we can reproduce in the future?

**Comment by tspiteri** [comment]
This is my code below. Should I include the code in a comment in f128.rs?

```rust
use rug::float::Constant;
use rug::{Assign, Float};

fn decimal_string(val: &Float, num_digits: usize) -> String {
    let (sign, mut digits, exp) = val.to_sign_string_exp(10, Some(num_digits));
    if let Some(exp) = exp {
        if exp <= 0 {
            let one_minus_exp = usize::try_from(1 - exp).unwrap();
            digits.insert_str(0, &"0".repeat(one_minus_exp));
            digits.insert(1, '.');
        } else {
            let exp = usize::try_from(exp).unwrap();
            if exp >= num_digits {
                digits.push_str(&"0".repeat(exp - num_digits));
            } else {
                digits.insert(exp, '.');
            }
        }
    }
    if sign {
        digits.insert(0, '-');
    }
    digits
}

fn float<T>(t: T) -> Float
where
    Float: Assign<T>,
{
    Float::with_val(1000, t)
}

fn print(name: &str, f: Float) {
    let s = decimal_string(&f, 60);
    assert_eq!(
        Float::with_val(113, Float::parse(&s).unwrap()),
        Float::with_val(113, &f)
    );
    println!("pub const {name}: f128 = {s}_f128;");
}

fn main() {
    print("PI", float(Constant::Pi));
    print("TAU", float(Constant::Pi) * 2);
    print("PHI", float(1.25).sqrt() + 0.5);
    print("EGAMMA", float(Constant::Euler));
    print("FRAC_PI_2", float(Constant::Pi) / 2);
    print("FRAC_PI_3", float(Constant::Pi) / 3);
    print("FRAC_PI_4", float(Constant::Pi) / 4);
    print("FRAC_PI_6", float(Constant::Pi) / 6);
    print("FRAC_PI_8", float(Constant::Pi) / 8);
    print("FRAC_1_PI", float(Constant::Pi).recip());
    print("FRAC_1_SQRT_PI", float(Constant::Pi).sqrt().recip());
    print("FRAC_2_PI", 2 / float(Constant::Pi));
    print("FRAC_2_SQRT_PI", 2 / float(Constant::Pi).sqrt());
    print("SQRT_2", float(2).sqrt());
    print("FRAC_1_SQRT_2", float(0.5).sqrt());
    print("SQRT_3", float(3).sqrt());
    print("FRAC_1_SQRT_3", float(3).sqrt().recip());
    print("E", float(1).exp());
    print("LOG2_10", float(10).log2());
    print("LOG2_E", float(1).exp().log2());
    print("LOG10_2", float(2).log10());
    print("LOG10_E", float(1).exp().log10());
    print("LN_2", float(2).ln());
    print("LN_10", float(10).ln());
}
```

**Comment by tgross35** [comment]
No this is fine, just wanted a reference in case anybody wants to replicate or add more in the future 👍

File: library/core/src/num/f128.rs [hunk 2]
```diff
@@ -15,13 +15,204 @@ use crate::mem;
...
+    pub const MIN_POSITIVE: f128 = 3.36210314311209350626267781732175260e-4932_f128;
+    /// Largest finite `f128` value.
+    ///
+    /// Equal to
+    /// (1&nbsp;&minus;&nbsp;2<sup>&minus;[`MANTISSA_DIGITS`]</sup>)&nbsp;2<sup>[`MAX_EXP`]</sup>.
+    ///
+    /// [`MANTISSA_DIGITS`]: f128::MANTISSA_DIGITS
+    /// [`MAX_EXP`]: f128::MAX_EXP
+    #[unstable(feature = "f128", issue = "116909")]
+    pub const MAX: f128 = 1.18973149535723176508575932662800701e+4932_f128;
```
Comment (tgross35) [inline library/core/src/num/f128.rs:186]: Minor style nit, would be good to consistently keep a line between the constant and docs (I know this was just copied from the existing floats)

**Review by tgross35 — APPROVED** [review]
Okay, these all look good to me. I double checked the associated values against http://weitz.de/ieee/ and the math constants against what Julia says https://github.com/tgross35/apfloat-consts/blob/49cb0d131ea2d881205852c31ba87236fdba9749/output.rs.

Honestly it would be better to just generate bit patterns for the `MIN`/`MAX`/`EPSILON`/etc values rather than having a tough to read literal, but that's not something to do now.

I can't r+ so somebody else has to. Thanks for the patch!

**Pull request merged**